tests/integration/bluez_controller.py

```diff
@@ -0,0 +1,165 @@
+
+
+@contextlib.asynccontextmanager
+async def wait_for_new_adapter() -> (
+    AsyncGenerator[tuple[MessageBus, asyncio.Future[str]], None]
+):
+    """
+    Connect to D-Bus and wait for a new Bluetooth adapter to be added.
+
+    Yields a Future that will be resolved with the adapter path when
+    a new adapter is detected via InterfacesAdded signal.
+    """
+    async with open_message_bus() as bus:
+        loop = asyncio.get_running_loop()
+        adapter_path_future: asyncio.Future[str] = loop.create_future()
+
+        def _on_interfaces_added(message: Message):
+            if message.message_type != MessageType.SIGNAL:
+                return
+            if message.member != "InterfacesAdded":
+                return
+
+            obj_path, ifaces = message.body
+            adapter = ifaces.get(defs.ADAPTER_INTERFACE)
+            if not adapter:
+                return
+
+            # New adapter found, return its path via the future
+            if not adapter_path_future.done():
+                adapter_path_future.set_result(obj_path)
+
+        bus.add_message_handler(_on_interfaces_added)
+
+        try:
+            # Subscribe to InterfacesAdded signals
+            reply = await add_match(
+                bus,
+                MatchRules(
+                    interface=defs.OBJECT_MANAGER_INTERFACE,
+                    member="InterfacesAdded",
+                    arg0path="/org/bluez/",
+                ),
+            )
+            assert_reply(reply)
+
+            # Yield the future and bus for the caller to use
+            yield bus, adapter_path_future
+        finally:
+            bus.remove_message_handler(_on_interfaces_added)
+
+
+@contextlib.asynccontextmanager
+async def open_bluez_bluetooth_controller_link(
+    hci_transport_name: str,
+) -> AsyncGenerator[LocalLink, None]:
+    """
+    Open a local link (virtual RF connection) to a bumble Bluetooth
+    controller that is connected to BlueZ.
+    """
+    async with wait_for_new_adapter() as (bus, adapter_path_future):
+        # Open a HCI transport connected to the OS
+        async with await open_transport(hci_transport_name) as hci_transport:
+            # Local link to create virtual RF connection between multiple Bluetooth Controllers
+            link = LocalLink()
+
+            # Bluetooth controller that BlueZ can connect to.
+            # (This will register itself to the link.)
+            Controller(
+                "BLEAK-TEST-BLUEZ",
+                host_source=hci_transport.source,
+                host_sink=hci_transport.sink,
+                link=link,
+            )
+
+            # Wait up to 5 seconds for the new adapter to appear via InterfacesAdded
+            adapter_path = await asyncio.wait_for(adapter_path_future, timeout=5.0)
+            logging.info(f"New adapter detected at {adapter_path}")
+
+            # Ensure controller is powered on
+            await power_on_controller(bus, adapter_path)
```

Comment:
Eventually, we will want to do something to ensure that Bleak uses this adapter in cases where there is more than one adapter present. But we should save that for a later change.